sdk/nodejs/runtime/closure/createClosure.ts

```diff
@@ -840,114 +871,94 @@ function getOrCreateEntry(
+            entry.function = await analyzeFunctionInfoAsync(obj, context, serialize, logInfo);
         }
-        else if (resource.Output.isInstance(obj)) {
-            // captures the frames up to this point. so we can give a good message if we
-            // fail when we resume serializing this promise.
-            const framesCopy = context.frames.slice();
-
-            // Push handling this output to the async work queue.  It will be processed in batches
-            // after we've walked as much of the graph synchronously as possible.
-            context.asyncWorkQueue.push(async () => {
-                const val = await obj.promise();
-                const oldFrames = context.frames;
-                context.frames = framesCopy;
-                entry.output = getOrCreateEntry(new SerializedOutput(val), undefined, context, serialize, logInfo);
-                context.frames = oldFrames;
-            });
+        else if (await isOutputAsync(obj)) {
+            entry.output = await createOutputEntryAsync(obj);
         }
         else if (obj instanceof Promise) {
-            // captures the frames up to this point. so we can give a good message if we
-            // fail when we resume serializing this promise.
-            const framesCopy = context.frames.slice();
-
-            // Push handling this promise to the async work queue.  It will be processed in batches
-            // after we've walked as much of the graph synchronously as possible.
-            context.asyncWorkQueue.push(async () => {
-                const val = await obj;
-                const oldFrames = context.frames;
-                context.frames = framesCopy;
-                entry.promise = getOrCreateEntry(val, undefined, context, serialize, logInfo);
-                context.frames = oldFrames;
-            });
+            const val = await obj;
+            entry.promise = await getOrCreateEntryAsync(val, undefined, context, serialize, logInfo);
         }
         else if (obj instanceof Array) {
-            // Recursively serialize elements of an array. Note: we use getOwnPropertyNames as the array
-            // may be sparse and we want to properly respect that when serializing.
+            // Recursively serialize elements of an array. Note: we use getOwnPropertyNames as the
+            // array may be sparse and we want to properly respect that when serializing.
             entry.array = [];
-            for (const key of Object.getOwnPropertyNames(obj)) {
-                if (key !== "length" && obj.hasOwnProperty(key)) {
-                    entry.array[<any>key] = getOrCreateEntry(
-                        obj[<any>key], undefined, context, serialize, logInfo);
+            for (const descriptor of await getOwnPropertyDescriptors(obj)) {
+                if (descriptor.name !== undefined &&
+                    descriptor.name !== "length") {
+
+                    entry.array[<any>descriptor.name] = await getOrCreateEntryAsync(
```

Comment:
`<string>`?

Reply:
descriptor.name is already a string.  so that wouldn't help.  the issue here is that *JS* allows and completely expects indexing into an array with a string-name (i.e. "1" or "57" etc.).  These is actuall indexing into the 1th and 57th positions appropriately.  However, TS took a more strict stance of going "you're probably doing something wrong if you do that, so i'll complain".  So this is to make TS ok since in this case (reflection) it's totally ok to index in with strings.